### ISSUE TITLE
jScroll doesn't seem to work ok with a page displaying results in a table

### ISSUE THREAD

**Issue by darkred** [pklauzinski/jscroll#60]
I've been trying to make this userscript in order to use jScroll it in thepiratebay 
(in the tables for Browse and Search results)

``` javascript
$('#searchResult').jscroll({    
    nextSelector: $('a').has('img[src="https://piratebay.org/img/next.gif"]'),
    contentSelector: 'tr',
    callback: function() {$('tr.header').remove();}
});
```

Test link: https://thepiratebay.gd/browse/401
_Note:the `nextSelector` is the link that appears with the ">" arrow icon._
<br/>
Unfortunately I got these problems:
1. jScroll **only appends (the cells of) page 2** and not the others too.
   Below is the HTML of the `a` tag of the "Next" link 
   _(FWIW, the  href value is a relative link)_
   
   ``` javascript
   <a href="/browse/401/1/3">
     <img src="https://piratebay.org/img/next.gif" alt="Next" border="0"></img>
   </a>
   ```
   
   _And, as you can see, the "Next" `<a>` anchor tag has no `id` or `class` 
   so I had to use a jQuery selector in `nextSelector` in order to select it._
2. **The rows from page 2 aren't properly aligned** to the table from page 1.  ([screenshot](http://i.imgur.com/6j59LME.jpg))
   Isn't the `contentSelector` right?

**Comment by pklauzinski** [comment]
@darkred - For `nextSelector` you can use the following string instead of creating a jQuery object:

```js
nextSelector: 'a:has(img[src="https://piratebay.org/img/next.gif"])'
```

Let me take a look into the other issues and I will get back to you.

**Comment by pklauzinski** [comment]
@darkred - Unfortunately, jScroll isn't currently designed to work within tables. It will add `<div>` elements within your table elements, creating invalid HTML. I would say your best option at this time is to call jScroll on `#main-content` and load each subsequent `<table>` from the jScroll payload underneath one another.

I do see the utility in having jScroll work within tables, however; so I will mark this as an enhancement.

**Comment by darkred** [comment]
Thanks a lot for your replies.

> For `nextSelector` you can use the following string instead of creating a jQuery object:
> 
> <div class="highlight highlight-js"><pre>nextSelector<span class="pl-k">:</span> <span class="pl-s"><span class="pl-pds">'</span>a:has(img[src="https://piratebay.org/img/next.gif"])<span class="pl-pds">'</span></span></pre></div>

Unfortunately the [relational pseudo-class](http://dev.w3.org/csswg/selectors-4/#relational) `:has()` (a Level 4 selector) is still not supported by FIrefox 39 (nor in Chrome 43).
_[Browser CSS-Selector-Test ](http://css4-selectors.com/browser-selector-test/)_


Your other suggestion (about calling jScroll on `#main-content` and using `table` as `contentSelector`) is indeed the best current option. Thanks!

**Comment by pklauzinski** [comment]
`:has()` is not supported natively by all browsers, but it is a jQuery selector and is normalized for unsupported browsers.

See: http://api.jquery.com/has-selector/

**Comment by darkred** [comment]
I thought that it wasn't a jQuery selector but a CSS one, 
because you wrote it without a dollar symbol in front 
i.e. not `nextSelector: $('a:has(img[src="https://piratebay.org/img/next.gif"])')`

But, unfortunately it doesn't work either way: it only appends page 2.
Please try it yourself:

```javascript
// ==UserScript==
// @name        thepiratebay jScroll
// @include     https://thepiratebay.*/browse/*
// @version     1
// @grant       none
// @require     http://code.jquery.com/jquery-2.1.4.min.js
// @require     https://cdn.rawgit.com/pklauzinski/jscroll/master/jquery.jscroll.min.js
// ==/UserScript==

$('#main-content').jscroll({    
    nextSelector: 'a:has(img[src="https://piratebay.org/img/next.gif"])',
    contentSelector: 'table'
});
```

in the end of the table (appended page 2) the last line which contains the "Next" link is missing (though it's contained inside the `table` element)

**Comment by mikila85** [comment]
is this going to be solved anytime soon?

**Comment by kmkroski** [comment]
I managed to get this working by using the callback. Basically, it grabs the new rows and new pagination from the last 'jscroll-added' div, replaces the old pagination, and appends the new rows to the table's tbody. Overly commented code below:

```
callback: function () {
    var newRows = $('.jscroll-added:last table tbody').html(),  // Get new rows
        newPage = $('.jscroll-added:last .pagination').html();  // Get new pagination
    $('.jscroll-added:last').remove();      // Remove the newly added 'jscroll-added' div
    $('.pagination').html(newPage);         // Replace existing pagination with newPage
    $('table > tbody').append(newRows);     // Append the new rows to the table's tbody
}
```

**Comment by darkred** [comment]
That's it! Thank you, @kmkroski

**Comment by wezzix** [comment]
I tried kmkroski's solution, but found two problems.
First, the next button only worked once - probably has something to do with copying html instead of dom objects with attached events. Solution: Replace pagination element.
Second, on the last page, there would be an additional table which was not merged. Solution: Use a selector for the table and use :first and :last.
Working solution, assuming table class="my-list":
```javascript
callback: function () {
    // Append new results to existing table
    $('.my-list:last > tbody tr').detach().appendTo($('.my-list:first > tbody'));
    // Replace existing pagination with new while keeping attached events
    $('.pager:first').replaceWith($('.pager:last'));
    // Remove the newly added 'jscroll-added' div, as well as the remaining table on the last page
    $('.jscroll-added,.my-list:nth(1)').remove();
}
```

**Comment by BENRAHAL-Aicha** [comment]
i tried @kmkroski 's solution as well as @wezzix 's one , and it didn't work for me at first , but turned out i had to fix a height for the div that contains the table : height: 100vh;
  overflow-y: auto; for example ,
and js code is like :  $(document).ready(function() {
      // alert('test');
             $('.ma-pagination').hide();
                // alert($('span[aria-current="page"] + a').attr('href'));
                console.log('span[aria-current="page"] + a',$('span[aria-current="page"] + a').attr('href'));
            $('.infinite-scroll').jscroll({
                        debug: true,
                        autoTrigger: true,
                        loadingHtml: '<p>loading...</p>',
                        padding: 0 ,
                        nextSelector: 'span[aria-current="page"] + a',
                        contentSelector: '.infinite-scroll',
                        callback: function () {
     // Append the new rows to the table's tbody
    $('.my-list:last > tbody tr').detach().appendTo($('.my-list:first > tbody'));
    // Replace existing pagination with new while keeping attached events
    $('.pagination:first').replaceWith($('.pagination:last'));
    // Remove the newly added 'jscroll-added' div, as well as the remaining unwrapped table on the last page
    $('.jscroll-added,.my-list:nth(1)').remove();


}
                    });
                
        })
